File: Reqnroll/Chapter17/17.2-InPlacePersonaPattern/Before/WIMP.Specs/StepDefinitions/ReportingStepDefinitions.cs
using DiffPlex.Renderer;

using Reqnroll;

using WIMP.App.Models;
using WIMP.Specs.Drivers;
using WIMP.Specs.Support;

namespace WIMP.Specs.StepDefinitions;

[Binding]
public class ReportingStepDefinitions(
    ReportingApiDriver reportingApiDriver,
    BackdoorApiDriver backdoorApiDriver)
{
    private SalesReport? generatedReport;

    [Given("a week of operations with these pizza sales:")]
    public async Task GivenAWeekOfOperationsWithThesePizzaSales(DataTable salesTable)
    {
        await backdoorApiDriver.PrepareSalesTraffic(salesTable).Execute();
    }

    [When("the sales report is requested for the week beginning {DateOnly}")]
    public async Task WhenTheSalesReportIsRequestedForTheWeekBeginning(DateOnly startDay)
    {
        generatedReport = await reportingApiDriver.GenerateSalesReport(startDay).Execute();
    }

    [Then("the report should show:")]
    public void ThenTheReportShouldShow(string expectedReport)
    {
        string actualReport = ReportPrinter.PrintReport(generatedReport ?? throw new InvalidOperationException("Report is not generated."));
        string diff = UnidiffRenderer.GenerateUnidiff(
            expectedReport, actualReport, "expected", "actual");
        Assert.IsTrue(string.IsNullOrEmpty(diff),
            "The report is different from the expected" + Environment.NewLine + diff);
    }

    [When("the ingredient usage report is requested for the week beginning {DateOnly}")]
    public void WhenTheIngredientUsageReportIsRequestedForTheWeekBeginning(DateOnly startDay)
    {
        // ingredient usage report generation is not part of this sample 
    }

    [Then("the ingredient usage report should show:")]
    public void ThenTheIngredientUsageReportShouldShow(string expectedReport)
    {
        // ingredient usage report generation is not part of this sample
    }
}

File: CucumberJVM/Chapter17/17.2-InPlacePersonaPattern/Before/src/test/java/com/wimp/app/specs/stepdefinitions/ReportingStepDefinitions.java
package com.wimp.app.specs.stepdefinitions;

import com.wimp.app.models.SalesReport;
import com.wimp.app.specs.drivers.BackdoorApiDriver;
import com.wimp.app.specs.drivers.ReportingApiDriver;
import com.wimp.app.specs.support.ReportPrinter;
import io.cucumber.datatable.DataTable;
import io.cucumber.java.en.*;

import java.time.LocalDate;
import java.util.*;

import static org.assertj.core.api.Assertions.assertThat;

public class ReportingStepDefinitions {
    private final BackdoorApiDriver backdoorApiDriver;
    private final ReportingApiDriver reportingApiDriver;
    private SalesReport generatedReport;

    public ReportingStepDefinitions(BackdoorApiDriver backdoorApiDriver, ReportingApiDriver reportingApiDriver) {
        this.backdoorApiDriver = backdoorApiDriver;
        this.reportingApiDriver = reportingApiDriver;
    }

    @Given("a week of operations with these pizza sales:")
    public void aWeekOfOperationsWithThesePizzaSales(DataTable salesTraffic) throws Exception {
        backdoorApiDriver.prepareSalesTraffic(salesTraffic).execute();
    }

    @When("the sales report is requested for the week beginning {date}")
    public void theSalesReportIsRequestedForTheWeekBeginning(LocalDate startDay) throws Exception {
        generatedReport = reportingApiDriver.generateSalesReport(startDay).execute();
    }

    @Then("the report should show:")
    public void theReportShouldShow(String expectedReport) {
        assertThat(ReportPrinter.printReport(generatedReport).trim().lines().toList())
            .as("The printed report did not match the expected output.")
            .containsExactlyElementsOf(expectedReport.trim().lines().toList());
    }

    @When("the ingredient usage report is requested for the week beginning {date}")
    public void theIngredientUsageReportIsRequestedForTheWeekBeginning(LocalDate startDay) {
      // ingredient usage report generation is not part of this sample
    }

    @Then("the ingredient usage report should show:")
    public void theIngredientUsageReportShouldShow(String expectedReport) {
      // ingredient usage report generation is not part of this sample
    }
}
